### PR TITLE
Vod support

### PR DESCRIPTION
**Pull request by Excellify**

### CODE REVIEW

File: src/site/twitch.tv/modules/vod-chat/VodChat.vue [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	const rootNode = props.list.domNodes.root;
+	if (!rootNode) return;
+
+	rootNode.classList.add("seventv-vod-chat-list");
+
+	containerEl.value = rootNode as HTMLElement;
+
+	const video = props.controller.component.videoData;
+	store.setChannel({
+		id: video.owner.id,
```
Comment (AnatoleAM) [inline src/site/twitch.tv/modules/vod-chat/VodChat.vue:73]: `video.owner` is sometimes null and this crashes the page

File: src/site/twitch.tv/modules/chat-data/ChatDataModule.vue [hunk 2]
```diff
@@ -17,6 +18,11 @@ const { channel } = storeToRefs(useStore());
 const messages = useChatMessages();
 const emotes = useChatEmotes();
 
+const { markAsReady } = declareModule("chat-data", {
+	name: "Chat Data",
+	depends_on: [],
+});
+
```
Comment (AnatoleAM) [inline src/site/twitch.tv/modules/chat-data/ChatDataModule.vue:25]: Why? This appears to serve no purpose

File: src/site/twitch.tv/modules/vod-chat/VodChat.vue [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+						@timeout="c.props.onDeleteComment(comment.comment)"
+					/>
+				</div>
+			</div>
+		</UiScrollable>
+		<!-- New Messages during Scrolling Pause -->
+		<div v-if="scroller.paused" class="seventv-message-buffer-notice" @click="scroller.unpause">
+			<span> Resume auto scroll </span>
+		</div>
+	</Teleport>
```
Comment (AnatoleAM) [inline src/site/twitch.tv/modules/vod-chat/VodChat.vue:20]: This shouldn't need to be a thing. Adapt the vod message type to work with the existing chatlist renderer

File: src/site/twitch.tv/modules/vod-chat/VodMessage.vue [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+	(event: "seek"): void;
+}>();
+
+const vodMsg = toRef(props, "vodMsg");
+
+const msg = new ChatMessage(props.vodMsg.comment.id);
+msg.setAuthor({
+	id: props.vodMsg.author.id,
+	username: props.vodMsg.author.name,
+	displayName: props.vodMsg.author.displayName,
```
Comment (AnatoleAM) [inline src/site/twitch.tv/modules/vod-chat/VodMessage.vue:47]: I am unable to comprehend why this is a thing? Why all this duplicate code when you could've simply adapted twitch's vod message type to fit with the existing tokenizer.

File: src/types/module.d.ts [hunk 5]
```diff
@@ -1,13 +1,17 @@
...
 import type ChatInputModuleVue from "@/site/twitch.tv/modules/chat-input/ChatInputModule.vue";
 import type ChatModuleVue from "@/site/twitch.tv/modules/chat/ChatModule.vue";
 import type EmoteMenuModuleVue from "@/site/twitch.tv/modules/emote-menu/EmoteMenuModule.vue";
 import type SettingsModuleVue from "@/site/twitch.tv/modules/settings/SettingsModule.vue";
+import type ChatVodModuleVue from "@/site/twitch.tv/modules/vod-chat/VodModule.vue";
 
 declare type ModuleID = keyof ModuleComponentMap;
 
 declare type ModuleComponentMap = {
+	"chat-data": typeof ChatDataModuleVue;
```
Comment (AnatoleAM) [inline src/types/module.d.ts:11]: Not sure what is the point of splitting "ChatData" into a new module

**Review by AnatoleAM — CHANGES_REQUESTED** [review]
General issue: doesn't load during regular navigation, need to refresh the entire page

**Comment by AnatoleAM** [comment]
tl;dr

- take out all the extra components you made
- instead, adapt the data from "comments" so they can be passed through the `onMessage()` in `ChatList.vue`
- then just modify the teleport logic so it fits onto the vod container

It should be possible to reduce this PR to like <200 lines of diff

File: src/types/module.d.ts (hunk 5 above)
Comment (Excellify) [reply src/types/module.d.ts:11]: Because chatController, chatList and chatRoom are not components when viewing a vod. So the chat-module dosnt load at all

File: src/site/twitch.tv/modules/vod-chat/VodMessage.vue (hunk 4 above)
Comment (Excellify) [reply src/site/twitch.tv/modules/vod-chat/VodMessage.vue:47]: Wdym, im using the tokenizer, but i need to build the message body from the msg tokens?

File: src/types/module.d.ts (hunk 5 above)
Comment (Excellify) [reply src/types/module.d.ts:11]: So that emotes can be fetched in places where the components that the chat module uses are unavailable

File: src/site/twitch.tv/modules/vod-chat/VodMessage.vue (hunk 4 above)
Comment (AnatoleAM) [reply src/site/twitch.tv/modules/vod-chat/VodMessage.vue:47]: in what way is there any code reusing here?  all I see is the same stuff copy pasted, this should be a 100 line diff branch

File: src/types/module.d.ts (hunk 5 above)
Comment (AnatoleAM) [reply src/types/module.d.ts:11]: why not just change the condition in the chatmodule so it can match either scenario?

File: src/site/twitch.tv/modules/vod-chat/VodChat.vue [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+const comments = ref<Twitch.VideoChatCommentContext[]>([]);
+const isMod = ref(false);
+
+const c = ref(props.controller.component);
+
+definePropertyHook(props.controller.component, "props", {
+	value(v) {
+		comments.value = v.comments;
+		isMod.value = v.canCurrentUserDelete;
+		nextTick(scroller.scrollToLive);
```
Comment (AnatoleAM) [inline src/site/twitch.tv/modules/vod-chat/VodChat.vue:57]: Scrolling to live on every prop update will lead to excessive reflows

File: src/site/twitch.tv/modules/vod-chat/VodChat.vue [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+			id: video.owner.id,
+			username: video.owner.login,
+			display_name: video.owner.login,
+			loaded: false,
+		});
+	}
+});
+
+onMounted(() => {
+	chatProperties.twitchBadgeSets = props.controller.component.badgeSets;
```
Comment (AnatoleAM) [inline src/site/twitch.tv/modules/vod-chat/VodChat.vue:83]: There is no guarantee badges would be populated by mount time?

File: src/site/twitch.tv/modules/vod-chat/VodMessage.vue [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+<script setup lang="ts">
+import { computed, toRef } from "vue";
+import { convertCheerEmote, convertTwitchEmote } from "@/common/Transform";
+import { ChatMessage } from "@/common/chat/ChatMessage";
+import TwMore from "@/assets/svg/twitch/TwMore.vue";
+import { MessagePartType } from "../..";
+import UserMessage from "../chat/components/message/UserMessage.vue";
+
+const props = defineProps<{
+	vodMsg: Twitch.VideoChatCommentContext;
```
Comment (AnatoleAM) [inline src/site/twitch.tv/modules/vod-chat/VodMessage.vue:31]: This should be our own message instance type

File: src/site/twitch.tv/modules/vod-chat/VodMessage.vue [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+			msg.body += "*".repeat(e.originalText.length);
+			break;
+		}
+		case MessagePartType.LINK:
+		case MessagePartType.CLIPLINK:
+		case MessagePartType.VIDEOLINK:
+			msg.body += part.content.url;
+			break;
+	}
+}
```
Comment (AnatoleAM) [inline src/site/twitch.tv/modules/vod-chat/VodMessage.vue:117]: Move this to `VodChat.vue`

**Comment by AnatoleAM** [comment]
Superseded by #104
